src/main/java/org/candlepin/model/CertificateSerialCurator.java

```diff
@@ -52,17 +55,53 @@ private Date getExpiryRestriction() {
-        String hql = "DELETE FROM CertificateSerial c WHERE c.revoked = true" +
-            " AND c.expiration < :cutoff";
-        Query query = this.getEntityManager().createQuery(hql);
-        query.setParameter("cutoff", getExpiryRestriction());
-        return query.executeUpdate();
+        EntityManager entityManager = this.getEntityManager();
+
+        // Impl. note: Under normal circumstances, we should not have to worry about deleting revoked
+        // serials being referenced in any certificate table, because the corresponding certificate is
+        // always deleted when the serial is revoked. However, this avoids bugs such as 2229095, where bad
+        // data caused by accident/bug could make this query fail.
+        //
+        // Also, there doesn't seem to be a way to do this kind of query with JQPL, so we're using native
+        // SQL, and MariaDB does not like table aliases in delete statements, so we have to do this
+        // in 2 queries (select, then delete).
+        List<String> nativeSpaces = List.of(CertificateSerial.class.getName(),
+            SubscriptionsCertificate.class.getName(),
+            EntitlementCertificate.class.getName(),
+            CdnCertificate.class.getName(),
+            IdentityCertificate.class.getName(),
+            ContentAccessCertificate.class.getName(),
+            UeberCertificate.class.getName());
+        String fetchSQL = "SELECT cs.id FROM cp_cert_serial cs WHERE cs.revoked = true " +
+            "AND cs.expiration < ? " +
+            "AND NOT EXISTS (SELECT 1 FROM cp_certificate subc WHERE subc.serial_id = cs.id) " +
+            "AND NOT EXISTS (SELECT 1 FROM cp_ent_certificate entc WHERE entc.serial_id = cs.id) " +
+            "AND NOT EXISTS (SELECT 1 FROM cp_cdn_certificate cdnc  WHERE cdnc.serial_id = cs.id) " +
+            "AND NOT EXISTS (SELECT 1 FROM cp_id_cert idc WHERE idc.serial_id = cs.id) " +
+            "AND NOT EXISTS (SELECT 1 FROM cp_cont_access_cert scac  WHERE scac.serial_id = cs.id) " +
+            "AND NOT EXISTS (SELECT 1 FROM cp_anonymous_certificates anonc  WHERE anonc.serial_id = cs.id) " +
+            "AND NOT EXISTS (SELECT 1 FROM cp_ueber_cert uebc WHERE uebc.serial_id = cs.id);";
+        List<Long> serials = entityManager.createNativeQuery(fetchSQL)
+            .setHint(QueryHints.NATIVE_SPACES, nativeSpaces)
+            .setParameter(1, getExpiryRestriction())
+            .getResultList();
+
+        String deleteSQL = "DELETE FROM cp_cert_serial WHERE id IN (:serials_to_delete)";
+        nativeSpaces = List.of(CertificateSerial.class.getName());
+        int deleted = 0;
+        for (List<Long> serialsToDeleteBlock : this.partition(serials)) {
+            deleted += entityManager.createNativeQuery(deleteSQL)
+                .setHint(QueryHints.NATIVE_SPACES, nativeSpaces)
+                .setParameter("serials_to_delete", serialsToDeleteBlock)
+                .executeUpdate();
```

Comment:
@Ceiu Is this what you had in mind in trying to avoid any kind of fk violation when the job runs? I am a bit conflicted with the potential performance hit here from subqeries and splitting into 2 queries, if it is really worth doing all this for avoiding any fk violation from rogue bad data.

Reply:
There's a lot to consider here:

- When it comes to certificates, the model is extremely silly. Having everything broken up the way it is comes down to a Hibernate-ism where the ownership of the joins is on what I would argue is the wrong side of the relationship. For instance certificates don't have entitlements, entitlements have certificates. As a direct result of this, doing any kind of cleanup or sanitation gets clunky and tedious. Worse, because the *cert* tables differ but all link back to a single table for the serials, we get into this situation where we have some subset of the certs which aren't revocable in terms of logic, but *can* be flagged as revoked in the data. Because of this desync possibility, IMO our logic needs to be solid enough that valid data in the database does not cause issues.
- It may or may not be more efficient to left join on the other tables instead of a correlated subquery. Given how simple those subqueries are, and how it basically looks like a join anyway, I'm betting the planner for PostgreSQL and MariaDB will end up doing nearly the same thing anyway. Worth looking into if performance is a concern with this.
- In terms of performance, how frequently is this going to be run? If this is part of the cleanup job that runs once/day at most, accuracy is more important to me than raw performance so long as that performance hit doesn't have knock-on effects elsewhere, and doesn't turn it into bubble-sort levels of bad. If it runs in under 60s as a periodic background task and isn't causing resource contention issues elsewhere, I'm not too concerned about it.
- In terms of certificates and how the model works logically, do we have any cases where we want to keep a cert around when we're deleting its serial? Putting aside the issue of non-revocable certs for a moment, it really feels like our model is backward here (again) -- we manage serials and certs are optional as far as our logic is concerned, so why is cert the parent and serial the child in this case? It prevents us from doing things like delete-cascades or any other method that would be less hands-on in terms of juggling both the cert and the serial. As it is now, we treat them both as related first-class objects rather than the parent-child relationship it actually is. Granted, this is starting to lean on a larger refactor a bit, but if we're starting to even consider shortcuts in terms of how we manage this stuff because it's either too slow or too tedious to do it right, maybe it's time we face this problem head-on.

Reply:
I used correlated subqueries because I initially was trying to do this in a single DELETE query, and Mariadb/Postgres have different syntaxes on how to do deletes with joins. But since I turned it into a SELECT I can use ANSI SQL LEFT joins now.

Performance wise: I run the correlated-subquery version in Prod and it took 2.5 minutes (that is, just for the SELECT, not including the subsequent DELETE with a large IN clause), while the LEFT JOIN version of the query takes 2-3 seconds. So we should be good here with this.

Reply:
Regarding the model situation (not something to deal with in this PR):

Logically, we never want to keep a cert after we deleted the serial.
I suspect the current model is the way it is, because the serial needs to be kept around for CRL purposes. 

I was looking into if we could remove the serial entirely from the subscription certificate entity (this is what the product certificates do, they don't actually have a serial), but it's probably not a good idea, because while subscription certificates are not supposed to be revocable, they still likely need the expiration data (which is part of the serial entity). I think satellite wants it (either they're using it, or they wanted to use it to warn users their manifests are about to expire).

So distilling the requirements down:
- We never want to keep a cert after we deleted the serial.
- We want some certs to be revocable, and some not.
- We want all certs to be expirable.
- For revocable certs, we need to track a state change from unrevoked to revoked.

If I would do this from scratch, this would be two separate entities and both of them contain all the fields of the current 2 entities (cert+serial). One of them would be mapped to cp_revocable_certificate, and the other to a cp_nonrevocable_certificate.

Entities that do not care about revocation (SubscriptionCertificate, ProductCertificate) would use the later, while all the others the former.

For the revocable ones, we wouldn't actually delete them as we do today. We would do a soft-delete, by simply changing the revoked field from false to true (the con is that this would use more space, by keeping all these revoked certs).

The CertificateCleanupJob would then just delete from cp_nonrevocable_certificate when they expire, and from the cp_revocable_certificate when they're both expired and revoked.